chain/chunks/src/logic.rs

```diff
@@ -0,0 +1,195 @@
+use std::sync::Arc;
+
+use near_chain::{validate::validate_chunk_proofs, Chain, ChainStore, RuntimeAdapter};
+use near_chunks_primitives::Error;
+use near_primitives::{
+    hash::CryptoHash,
+    merkle::{merklize, MerklePath},
+    receipt::Receipt,
+    sharding::{
+        EncodedShardChunk, PartialEncodedChunk, PartialEncodedChunkPart, PartialEncodedChunkV1,
+        PartialEncodedChunkV2, ReceiptProof, ShardChunk, ShardChunkHeader, ShardProof,
+    },
+    types::{AccountId, ShardId},
+};
+use tracing::log::{debug, error};
+
+#[derive(Clone)]
+pub struct ChunksLogic {
+    pub me: Option<AccountId>,
+    pub runtime_adapter: Arc<dyn RuntimeAdapter>,
+}
+
+impl ChunksLogic {
```

Comment:
@mzhangmzz Moving your comment from the forked repo PR here:

> I'm not sure if it simplifies the code by creating a new struct that includes AccountId and RuntimeAdapter. I would think just creating some static methods is enough and also easier to understand, especially that here the shared parameters are only two.

Reply:
I was debating between these too, but here is my rationale:

1. there are quite a few methods here, such as need_part, and they used to be instance functions that read pretty well, like self.need_part(prev block, the part). If we were to convert them to pure static functions, they would read like self.need_part(prev block, the part, me, runtime adapter). It makes the function call less clear what it is actually doing, whereas putting me and runtime_adapter into the struct makes it clearer that those are background context.
2. the Client also needs these two fields; it already has runtime_adapter but it doesn't have "me". The fact that Client has never needed "me" makes me wonder that maybe we should not just dump "me" as a field in there. If we just make it a field, then it makes it easier for future code to start depending on the "me" field which may not be desirable; keeping it in ChunksLogic however communicates the intent that the "me" field is only there for computing chunks-related logic.

Wdyt?

Reply:
Thanks for the detailed reasoning. I'm still not sure about it because 
1. There are also many methods here that don't fit very well as instance functions. For example, `cares_about_shard_this_or_next_epoch`, which takes another account_id as argument and doesn't use me at all, or `make_outgoing_receipts_proofs` or `create_partial_chunk` which doesn't use me at all. I think the only functions that read better as instance functions are `need_part` and `need_receipt`, and there are quite a few other functions that feel a bit awkward to be here. 
2. Good observation that "me" is not part of client. I thought very hard but I can't actually think of a particular reason for it not to be. And putting it inside chunk_logic doesn't help with the separation that much since any function can just access "me" through chunk_logic.

Reply:
Yeah, fair enough. Let me make these static functions then. Thanks!!

Reply:
Don't take this as a design advice for this specific PR, but I want to share some vague broad context about grouping stuff in Rust I myself converged to. 

I'd say in Rust bunching things up into structs for convenient access doesn't work well. So, the following setup is suspicions:


```rust
struct Parent {
    ctx: Arc<Ctx>,
    child: Child
}

struct Child {
    ctx: Arc<Ctx>
}

impl Child {
  fn foo(&self)
}
```

Here, ctx is aliased between parent and child, and in Rust that usually don't fly well. Rust prefers clear ownership structure: 


```rust
struct Parent {
    ctx: Arc<Ctx>,
    child: Child
}

struct Child {

}

impl Child {
  fn foo(&self, ctx: &Ctx)
}
````

What works for *logical* grouping is modules and top-level functions. Spitballing a bit, 


```
// chain/src/lib.rs
mod block;
mod chunk;

pub struct Chain {
    
}

pub enum Event { NewBlock(Block), NewChunk(Chunk) }

impl Chain {
  pub fn on_event(&mut self, event: Event) {
    match event {
      Event::NewBlock(block) => block::accept(self, block),
      Event::NewChunk(chunk) => chunk::accept(self, chunk),
    }
  }
}

// chain/src/block.rs
pub(super) fn accept(chain: &mut Chain, new_block: Block) {
}

fn helper(chain: &mut Chain, new_block: Block) {
}

// chain/src/chunk.rs
pub(super) fn accept(chain: &mut Chain, new_chunk: Chunk) {
}
```


In this setup, large contextual structs contain only "atomic" access functions as method, and the actual logic for semi-orthogonal bits of functionality is just top-level functions separated by module. 

If a particular processing logic grows to have a bunch of functions which thread through a bunch of contextual argument together, it's useful to bunch those up into a short-lived context object. Zooming in onto the block.rs, it could look like

```
// public intreface is still an fn
pub(supre) fn accept(chain: &mut Chain, new_block: Block) {
  Ctx { chain, new_block }.accept()
}

// Short-lived context object with a lifetime. 
struct Ctx<'a> {
  chain: &mut Chain,
  new_block: Block,
}

impl Ctx<'a> {
    fn accept(self)
    fn helper(&mut self)
}

```